### ISSUE TITLE
printf() will print one line twice in SGX mode

### ISSUE THREAD

**Issue by dashuaic** [oscarlab/graphene#2091]
## Description of the problem
The printf() will output one line before fork() twice. if add the fflush() after first printf(), will get the correct result. 
and if there is no fork(), will get correct result as well. it seems the fork() function reset the cache buffer offset. 

## Steps to reproduce

c source file 
> #include <sys/mman.h>
> #include <sys/types.h>
> #include <sys/wait.h>
> #include <stdio.h>
> #include <unistd.h>
> 
> int main(void) {
>  int shared = 0;
>         pid_t child;
>         int childstate;
>         printf("%d\n", shared);
>         //fflush(stdout);
>         if((child=fork())==0){
>                 exit(0);
>         }
>         waitpid (child, &childstate, 0);
>         printf("%d\n",shared);
> }
## Expected results
0
0
## Actual results
0
0
0

**Comment by dimakuv** [comment]
This is a known issue but we consider it not worth our time/energy to fix this. Please see these discussions:
- https://github.com/oscarlab/graphene/issues/1212
- https://github.com/oscarlab/graphene/pull/1423

TLDR: Glibc has two modes of STDOUT: TTY (app writes to terminal) and non-TTY (app writes to pipe/file). Graphene supports only one of them (non-TTY).

If you want to know the crazy details about this:
- https://eklitzke.org/stdout-buffering
- https://stackoverflow.com/questions/56419346/how-does-the-libc-function-isatty-work

If you want to fix this behavior of Glibc, simply use this trick in your manifest:
https://github.com/oscarlab/graphene/blob/master/LibOS/shim/test/ltp/manifest.template#L5-L6

Be aware! This is a very ugly feature of Glibc, and you should not rely on this behavior (i.e., please do not rely on the order of printed lines).

**Comment by boryspoplawski** [comment]
If you can modify the source code, just add `setbuf(stdout, NULL)` (or other `FILE*` you are using) early in the initialization (global constructor or just inside `main`).

**Comment by dashuaic** [comment]
@boryspoplawski yes, setbut(stdout, NULL) is a work around. also setlinebuf(stdout) can works as well. I think this is a little different with #1212. That issue was caused by close stdout early, but this issue the stdout is in there but print 2 times.

**Comment by boryspoplawski** [comment]
@dashuaic This is not a workaround, this is the correct solution. The following program will print `ABC\n` twice on normal Linux (without Graphene):
```
int main(void) {
    printf("asd\n");
    fork();
    return 0;
}
```
When you run it like that: `./a.out | cat`, or even `./a.out` if you remove the newline. This is the expected behavior.

**Comment by mkow** [comment]
@dashuaic Just as a note, the code you pasted is actually not correct, i.e. it can sometimes output three zeros even on Linux, e.g. when you redirect the output to a file:

```
~/test> ./test > output
~/test> cat output 
0
0
0
```

You should flush the buffers manually before doing fork(). Or better, you should never use fork, it's a low-level syscall not intended for "end-users" but rather language frameworks.

**Comment by mkow** [comment]
Oops, I just noticed that Borys already wrote the same. I had this GitHub page cached and didn't refresh it before answering, sorry.
